File: ui/src/core/bid-link.ts
/**
 * BidLink: handles bid:\\ protocol links in Markdown format.
 * Port of Python book_index_manager.bid_link
 *
 * Format: [Title](bid:\\ID)
 */

import type { IndexType } from '../types';
import { base36Encode, smartDecode, decodeIdString } from '../id';

export class BidLink {
    static readonly PROTOCOL = 'bid:\\\\';
    static readonly PREFIX = 'bid:\\\\';

    readonly title: string;
    readonly idStr: string;
    readonly idInt: bigint;

    private _type: IndexType | null = null;

    constructor(idVal: string | bigint, title: string = '') {
        this.title = title;

        if (typeof idVal === 'string') {
            if (idVal.startsWith(BidLink.PREFIX)) {
                idVal = idVal.slice(BidLink.PREFIX.length);
            }
            this.idStr = idVal;
            try {
                this.idInt = smartDecode(idVal);
            } catch {
                this.idInt = 0n;
            }
        } else {
            this.idInt = idVal;
            this.idStr = base36Encode(idVal);
        }

        if (this.idInt > 0n) {
            try {
                const components = decodeIdString(this.idStr);
                this._type = components.type;
            } catch {
                // ignore
            }
        }
    }

    get type(): IndexType | null {
        return this._type;
    }

    getIcon(): string {
        if (this._type === null) return '';
        if (this._type === 'book') return '📖 ';
        if (this._type === 'collection') return '📚 ';
        if (this._type === 'work') return '📜 ';
        return '';
    }

    render(withIcon: boolean = false): string {
        const icon = withIcon ? this.getIcon() : '';
        return `[${icon}${this.title}](${BidLink.PREFIX}${this.idStr})`;
    }

    static parseFromLink(markdownLink: string): BidLink | null {
        const match = markdownLink.match(/\[(.*?)\]\((.*?)\)/);
        if (match) {
            const title = match[1];
            const url = match[2];
            if (url.startsWith(BidLink.PREFIX)) {
                const idPart = url.slice(BidLink.PREFIX.length);
                return new BidLink(idPart, title);
            }
        }
        return null;
    }

    static isBidLink(url: string): boolean {
        return url.startsWith(BidLink.PREFIX);
    }
}

File: book_index_manager/bid_link.py
import re
from typing import Optional, Union
from .id_generator import BookIndexIdGenerator, BookIndexType, base36_encode, smart_decode


class BidLink:
    r"""
    Component for handling Bid (Book Index ID) links.
    Format: [Title](bid:\\ID)
    """

    PROTOCOL = "bid:\\\\"
    PREFIX = "bid:\\\\"

    def __init__(self, id_val: Union[int, str], title: str = ""):
        self.title = title
        if isinstance(id_val, str):
            if id_val.startswith(self.PREFIX):
                id_val = id_val[len(self.PREFIX):]
            self.id_str = id_val
            try:
                self.id_int = smart_decode(id_val)
            except ValueError:
                self.id_int = 0
        else:
            self.id_int = id_val
            self.id_str = base36_encode(id_val)

        self._type = None
        if self.id_int > 0:
            try:
                components = BookIndexIdGenerator.parse(self.id_int)
                self._type = components.type
            except Exception:
                pass

    @property
    def type(self) -> Optional[BookIndexType]:
        return self._type

    def get_icon(self) -> str:
        if self._type is None:
            return ""
        if self._type == BookIndexType.Book:
            return "📖 "
        elif self._type == BookIndexType.Collection:
            return "📚 "
        elif self._type == BookIndexType.Work:
            return "📜 "
        return ""

    def render(self, with_icon: bool = False) -> str:
        icon = self.get_icon() if with_icon else ""
        return f"[{icon}{self.title}]({self.PREFIX}{self.id_str})"

    @staticmethod
    def parse_from_link(markdown_link: str) -> Optional['BidLink']:
        match = re.search(r'\[(.*?)\]\((.*?)\)', markdown_link)
        if match:
            title = match.group(1)
            url = match.group(2)
            if url.startswith(BidLink.PREFIX):
                id_part = url[len(BidLink.PREFIX):]
                return BidLink(id_part, title)
        return None

    @staticmethod
    def is_bid_link(url: str) -> bool:
        return url.startswith(BidLink.PREFIX)
